### PR TITLE
Chore: Product name and type

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Changelog Description
Goal of this PR is to start using product naming in code base. Which means to use `productName` and `productType` instead of `subset` and `family` during create and publish phase.

## Additional info
First of all I'm 100% sure that I broke something, if you can, please do code review first before you try to run. If you find any issues please report as soon as possible, you can DM me with quick questions.

All changes should support compatibility with workfiles before this PR in all hosts.

All the time when a `family` did represent product type, it should be changed to `product type`. e.g. create plugin does have `product_type` instead of `family` and created instance has `productType` in data instead of `family`.

One of the goals was to also use `productType` instead of `family` during publishing and use only `families` for pyblish filtering. The key `family` key is filled by pyblish automaticlly to `"default"`, so it should not be an issue. **But** when we planned it, we didn't realize that sometimes `instance.data` are filled with raw data from scene, and in that case there might be `family` with a value that is not managed by us, speaking only about logic or workfiles with instances before this PR. I've changed that idea and all the time `productType` is changed during publishing, `family` is changed too.

## What I couldn't resolve
1. Json files in **Blender** and **Unreal**
Sometimes it looks like the json is storing data that are unused, and sometimes I couldn't find source of their data so I couldn't tell if/how should I modify them. Sometimes the data are stored unclearly under key that says nothing about the value, e.g. json has `"_id"` which in fact represents representation id. Some of the json files are storing `family` which should be probably changed to `product_type` but I didn't find where the value is used. etc.

2. Json files in **Maya**
This is just about looks publishing, needs validation that I didn't break anything.

3. **Resolve** and **Hiero**
I probably failed at backwards compatibility.

## Testing notes:
- For host testing go one by one.
- Go through code if you can and scan if you find invalid changes.
- Open host and try all create plugin and publish. Validated output if can be loaded.
- Open the host you're testing with workfile that does have instances before this PR and publish them.
- [x] 3dsMax
- [x] AfterEffects
- [x] Blender
- [x] Fusion
- [x] Hiero
- [x] Houdini
- [x] Maya
- [x] Nuke
- [x] Photoshop
- [x] Resolve
- [x] SubstancePainter
- [x] TrayPublisher - simple creators
- [x] TrayPublisher - editorial
- [x] TVpaint
- [x] Unreal
- [ ] ~~Celaction~~
- [ ] ~~Flame~~
- [ ] ~~Harmony~~

Note: This PR is clashing Hiero integration PR https://github.com/ynput/ayon-core/pull/87

### CODE REVIEW

**Comment by moonyuet** [comment]
Tested in Substance Painter, Looks good.
Tested in 3dsMax, so far so good,

**Comment by MustafaJafar** [comment]
Who could believe that `productName` exists in one of Houdini nodes 🥲.
Any ideas @moonyuet @antirotor @BigRoy 🥹 ?

image

**Comment by iLLiCiTiT** [comment]
> Who could believe that productName exists in one of Houdini nodes

Houdini does store `AYONProductName` from now, please try it out.

**Comment by MustafaJafar** [comment]
I think there's something broken in the `dynamic_data` 
Here's a test with the same Ayon settings.

<table>
<tr>
<td> This branch </td>
<td> Develop Branch </td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

**Comment by iLLiCiTiT** [comment]
> I think there's something broken in the dynamic_data
Here's a test with the same Ayon settings.

Please be more specific, I don't see anything broken, nor even know where to look.

**Comment by MustafaJafar** [comment]
> Please be more specific, I don't see anything broken, nor even know where to look.

The product name in the publisher doesn't follow the product name profile in my settings.

**Review by kalisp — APPROVED** [review]
Tested in PS, AE, Fusion.

**Comment by simonebarbieri** [comment]
Getting this when opening the creator in Blender:
```
Traceback (most recent call last):
  File "C:\Dev\Ynput\ayon-core\client\ayon_core\pipeline\create\context.py", line 2050, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "C:\Dev\Ynput\ayon-core\client\ayon_core\hosts\blender\plugins\create\create_workfile.py", line 46, in create
    product_name = self.get_product_name(
  File "C:\Dev\Ynput\ayon-core\client\ayon_core\pipeline\create\creator_plugins.py", line 549, in get_product_name
    return get_product_name(
  File "C:\Dev\Ynput\ayon-core\client\ayon_core\pipeline\create\product_name.py", line 135, in get_product_name
    asset_tasks = asset_doc.get("data", {}).get("tasks") or {}
AttributeError: 'str' object has no attribute 'get'
```

File: client/ayon_core/hosts/blender/plugins/publish/extract_blend.py [hunk 1]
```diff
@@ -20,9 +20,9 @@ def process(self, instance):
         # Define extract output file path
 
         stagingdir = self.staging_dir(instance)
-        asset_name = instance.data["assetEntity"]["name"]
-        subset = instance.data["subset"]
-        instance_name = f"{asset_name}_{subset}"
+        folder_name = instance.data["assetEntity"]["name"]
+        product_name = instance.data["prouctName"]
```
Comment (simonebarbieri) [inline client/ayon_core/hosts/blender/plugins/publish/extract_blend.py:24]:
```suggestion
        product_name = instance.data["productName"]
```

File: client/ayon_core/hosts/blender/plugins/publish/extract_blend_animation.py [hunk 2]
```diff
@@ -23,9 +23,9 @@ def process(self, instance):
         # Define extract output file path
 
         stagingdir = self.staging_dir(instance)
-        asset_name = instance.data["assetEntity"]["name"]
-        subset = instance.data["subset"]
-        instance_name = f"{asset_name}_{subset}"
+        folder_name = instance.data["assetEntity"]["name"]
+        product_name = instance.data["prouctName"]
```
Comment (simonebarbieri) [inline client/ayon_core/hosts/blender/plugins/publish/extract_blend_animation.py:27]:
```suggestion
        product_name = instance.data["productName"]
```

File: client/ayon_core/hosts/blender/plugins/publish/extract_camera_fbx.py [hunk 3]
```diff
@@ -20,9 +20,9 @@ def process(self, instance):
 
         # Define extract output file path
         stagingdir = self.staging_dir(instance)
-        asset_name = instance.data["assetEntity"]["name"]
-        subset = instance.data["subset"]
-        instance_name = f"{asset_name}_{subset}"
+        folder_name = instance.data["assetEntity"]["name"]
+        product_name = instance.data["prouctName"]
```
Comment (simonebarbieri) [inline client/ayon_core/hosts/blender/plugins/publish/extract_camera_fbx.py:24]:
```suggestion
        product_name = instance.data["productName"]
```

File: client/ayon_core/hosts/blender/plugins/publish/extract_fbx_animation.py [hunk 4]
```diff
@@ -145,9 +145,9 @@ def process(self, instance):
 
         root.select_set(True)
         armature.select_set(True)
-        asset_name = instance.data["assetEntity"]["name"]
-        subset = instance.data["subset"]
-        instance_name = f"{asset_name}_{subset}"
+        folder_name = instance.data["assetEntity"]["name"]
+        product_name = instance.data["prouctName"]
```
Comment (simonebarbieri) [inline client/ayon_core/hosts/blender/plugins/publish/extract_fbx_animation.py:149]:
```suggestion
        product_name = instance.data["productName"]
```

File: client/ayon_core/hosts/blender/plugins/publish/extract_playblast.py [hunk 5]
```diff
@@ -55,9 +55,9 @@ def process(self, instance):
 
         # get output path
         stagingdir = self.staging_dir(instance)
-        asset_name = instance.data["assetEntity"]["name"]
-        subset = instance.data["subset"]
-        filename = f"{asset_name}_{subset}"
+        folder_name = instance.data["assetEntity"]["name"]
+        product_name = instance.data["prouctName"]
```
Comment (simonebarbieri) [inline client/ayon_core/hosts/blender/plugins/publish/extract_playblast.py:59]:
```suggestion
        product_name = instance.data["productName"]
```

File: client/ayon_core/tools/publisher/control.py [hunk 6]
```diff
@@ -2096,7 +2097,7 @@ def get_subset_name(
         if instance_id:
             instance = self.instances[instance_id]
 
-        return creator.get_subset_name(
+        return creator.get_product_name(
```
Comment (MustafaJafar) [inline client/ayon_core/tools/publisher/control.py:2100]: I'm not sure how should be corrected. but, I think there's something in the arguments of this method `creator.get_product_name` call.
As, they don't match with this: 
https://github.com/ynput/ayon-core/blob/8dd9d25ac44ec58e1ca0e7a0214f846d92adf0c1/client/ayon_core/pipeline/create/product_name.py#L82-L93

**Review by MustafaJafar — APPROVED** [review]
I tested it in Houdini and it works. 
Few notes for reference. 
**Note 1:**
It'd be better to prefix all parameters as you and @BigRoy mentioned here https://github.com/ynput/ayon-core/pull/115#issuecomment-1962657403 . I think we can do it in another PR as the current status doesn't conflict with the implemented product types. 

**Note 2:**
I faced few bugs but I think they are not related to this PR as I was able to replicate them in develop branch.
- Publish Image sequence (for some reason, representation data are empty)
- Loading HDAs from Loader (where Loader fetches the wrong version)

**Comment by jakubjezek001** [comment]
**Hiero** and **Resolve** works!

**Review by moonyuet — APPROVED** [review]
Tested in 3dsmax and substance painter. So far so good.

File: client/ayon_core/tools/publisher/control.py (hunk 6 above)
Comment (iLLiCiTiT) [reply client/ayon_core/tools/publisher/control.py:2100]: Good point, changed order in the method to look more similar (not exactly the same).

**Comment by jakubjezek001** [comment]
Traypublisher's **Simple EDL editorial** works like a charm ;)

**Comment by jakubjezek001** [comment]
Traypublisher's **Simple Creators** works too.

**Comment by MustafaJafar** [comment]
Creator Tab in Publisher tool is able to get the product name correctly. 
Thanks.
image

**Review by simonebarbieri — APPROVED** [review]
Blender and Unreal works. There are a few issues in Unreal, but unrelated to this PR.

**Review by antirotor — APPROVED** [review]
I've hit issue when publishing renders from maya, but it doesn't seem to be connected to this PR. So fo me and Maya 👍🏻

File: client/ayon_core/hosts/unreal/api/plugin.py [hunk 7]
```diff
@@ -53,34 +53,36 @@ def cache_subsets(shared_data):
...
-
-            shared_data["unreal_cached_subsets"] = unreal_cached_subsets
-            shared_data["unreal_cached_legacy_subsets"] = (
-                unreal_cached_legacy_subsets
-            )
-        return shared_data
 
-    def create(self, subset_name, instance_data, pre_create_data):
+        if "unreal_cached_instances" in shared_data:
+            return
```
[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply client/ayon_core/hosts/unreal/api/plugin.py:54]: I'm sorry but I don't think it is necessary to return argument that is passed to the method. It's not used that way anywhere. I've removed `Returns` from all docstring.

**Pull request merged**